internals/overlord/pairingstate/manager.go

```diff
@@ -0,0 +1,312 @@
+
+// pairingstate manages client server pairing.
+package pairingstate
+
+import (
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/canonical/pebble/internals/overlord/state"
+	"github.com/canonical/pebble/internals/plan"
+)
+
+// timeAfterFunc allows faking time.AfterFunc.
+var timeAfterFunc = func(d time.Duration, f func()) Timer {
+	return time.AfterFunc(d, f)
+}
+
+// Timer is used so we can supply a fake timer during testing.
+type Timer interface {
+	Stop() bool
+	Reset(time.Duration) bool
+}
+
+// autoUsernameRangeLimit determines the maxmimum number suffix for
+// users auto-allocated by this package when a new certificate is
+// paired.
+const autoUsernameRangeLimit uint32 = 1000
+
+// pairedStateKey is the key to the paired state. If the paired state is true
+// least one client successfully paired with the server. The paired state
+// is significant for "single" pairing mode because once the first client
+// paired with the server (and paired state is set to true), no futher pairing
+// is allowed from that point in time (until the state is cleared).
+const pairedStateKey = "paired"
+
+// Mode controls the pairing policy of the pairing manager.
+type Mode string
+
+const (
+	// ModeUnset is the same as ModeDisabled, but this value prevents the
+	// plan from marshalling the Mode explicitly.
+	ModeUnset Mode = ""
+	// ModeDisabled means no pairing is possible
+	ModeDisabled Mode = "disabled"
+	// ModeSingle means only a single client can pair.
+	ModeSingle Mode = "single"
+	// ModeMultiple means multiple clients can pair.
+	ModeMultiple Mode = "multiple"
+)
+
+var _ plan.Section = (*PairingConfig)(nil)
+
+// PairingConfig contains the options exposed in the plan extension.
+type PairingConfig struct {
+	Mode Mode `yaml:"mode,omitempty"`
+}
+
+func (c *PairingConfig) Validate() error {
+	switch c.Mode {
+	case ModeUnset, ModeDisabled, ModeSingle, ModeMultiple:
+	default:
+		return fmt.Errorf("cannot support pairing mode %q: unknown mode", c.Mode)
+	}
+	return nil
+}
+
+// Implements the optional Zeroer interface as used by the YAML library
+// for deciding when to marshal a section or not.
+func (c *PairingConfig) IsZero() bool {
+	return c.Mode == ModeUnset
+}
+
+func (c *PairingConfig) Combine(other *PairingConfig) {
+	if other.Mode != ModeUnset {
+		c.Mode = other.Mode
+	}
+}
+
+type PairingManager struct {
+	state  *state.State
+	mu     sync.Mutex
+	config *PairingConfig
+	// paired is true if a previous pairing request succeeded.
+	paired bool
+	// enabled is true if the pairing window is enabled.
+	enabled bool
+	// timer controls the duration of the pairing window.
+	timer Timer
+}
+
+func NewManager(state *state.State) *PairingManager {
+	m := &PairingManager{
+		state: state,
+		config: &PairingConfig{
+			Mode: ModeUnset,
+		},
+	}
+
+	// Load the paired state at startup.
+	m.state.Lock()
+	defer m.state.Unlock()
+	m.state.Get(pairedStateKey, &m.paired)
```

Comment:
Let's please define a `type pairingState struct` here, as otherwise if we ever need to store a second field, we'll be stuffing it all in the general namespace of the global state as well.  The key used should follow along, and the type names as well.  I believe we have a "pairing-details", PairingDetails, etc convention, as one of the strong arms of this.  The project next door has some recent examples, I believe.